### PR TITLE
Task/frontend29 configure notifications

### PR DESCRIPTION
**Pull request by shirokuronemuri**

### CODE REVIEW

**Comment by shirokuronemuri** [comment]
[Configure notifications](https://trello.com/c/s7pWDKVC/29-configure-notifications)

File: frontend/src/components/common/header/header.module.scss [hunk 1]
```diff
@@ -166,6 +170,43 @@
     }
   }
 
+  .unread-mark {
+    position: absolute;
+    top: 0%;
```
Comment (kyrylo-burmelov) [inline frontend/src/components/common/header/header.module.scss]: Not a mistake, but you can use here `top: 0` without `%`

File: frontend/src/components/common/header/header.tsx [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+import { FC, NotificationsResponseDto } from 'common/types/types';
 import { Link } from 'react-router-dom';
 import { FormEvent, MouseEvent, RefObject } from 'react';
 import { AppRoute, IconName } from 'common/enums/enums';
-import { MenuOptions } from 'common/enums/components';
+import { IconColor, MenuOptions } from 'common/enums/components';
 import { Icon } from '../icon';
 
 import styles from './header.module.scss';
+import { NotificationDropdown } from 'components/notification-dropdown/notification-dropdown';
```
Comment (kyrylo-burmelov) [inline frontend/src/components/common/header/header.tsx]: import order, you have empty line before styles import, so this has to be above styles

File: frontend/src/components/common/header/header.tsx [hunk 3]
```diff
@@ -79,8 +92,21 @@ const Header: FC<HeaderProps> = ({
             <button className={styles['btn-go-stream']}>
               <Icon name={IconName.CAMERA} width="30" height="24" />
             </button>
-            <button className={styles['btn-notification']}>
-              <Icon name={IconName.BELL} width="24" height="27" />
+            <button className={styles['btn-notification']} onClick={handleClickNotificationsMenu}>
+              <>
+                <Icon color={IconColor.GRAY} name={IconName.ALARM} width="24" height="27" />
+                {haveUnreadNotifications && <div className={styles['unread-mark']}></div>}
```
Comment (kyrylo-burmelov) [inline frontend/src/components/common/header/header.tsx]: you can use self-closing tags in such cases

File: frontend/src/components/common/header/header.tsx [hunk 4]
```diff
@@ -79,8 +92,21 @@ const Header: FC<HeaderProps> = ({
...
+                {isNotificationsMenuOpen && (
+                  <div
+                    ref={notificationsMenuRef}
+                    className={`${styles['notifications-wrapper']} ${
+                      isNotificationsMenuOpen && styles['notifications-dropdown-open']
+                    }`}
```
Comment (kyrylo-burmelov) [inline frontend/src/components/common/header/header.tsx]: use clsx for managing classNames, see more here https://www.npmjs.com/package/clsx

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  onRead(id: string): void;
+};
+
+const Notification: FC<NotificationProps> = ({ notification, onRead }) => {
+  const handleNotificationRead = (): void => {
+    onRead(notification.id);
+  };
+
+  return (
+    <Link className={styles['link']} to={`/video/${notification.videoId}`}>
```
Comment (kyrylo-burmelov) [inline frontend/src/components/notification-dropdown/components/notification/notification.tsx]: Does it mean that every notification has link to a video page? That's incorrect

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      <div className={styles['notification']} onClick={handleNotificationRead}>
+        <div className={`${styles['status']} ${!notification.isViewed && styles['status-unread']}`}></div>
```
Comment (kyrylo-burmelov) [inline frontend/src/components/notification-dropdown/components/notification/notification.tsx]: again clsx

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        <img
+          src={notification.channelAvatar ?? DefaultUserAvatar}
+          className={styles['avatar']}
+          alt="channel-avatar"
+          height="20"
+          width="21"
+        />
+        <div className={styles['info']}>
+          <p className={styles['message']}>
+            {notification.username} has started streaming: {notification.videoName}
```
Comment (kyrylo-burmelov) [inline frontend/src/components/notification-dropdown/components/notification/notification.tsx]: Notification body must be generic

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+          </p>
+          <p className={styles['timestamp']}>{getHowLongAgoString(notification.createdAt)}</p>
+        </div>
+        <img src={notification.videoPreview} className={styles['thumbnail']} alt="channel-avatar" width="60" />
```
Comment (kyrylo-burmelov) [inline frontend/src/components/notification-dropdown/components/notification/notification.tsx]: Not all notifications would have an image, it has to be optional

**Review by kyrylo-burmelov — CHANGES_REQUESTED** [review]
Let me know if you have any questions

File: frontend/src/components/common/header/header.tsx (hunk 4 above)
Comment (kyrylo-burmelov) [reply frontend/src/components/common/header/header.tsx]:
```
className={clsx(styles['notifications-wrapper'], { ["is-active"]: true })}
```

File: frontend/src/components/common/header/header.module.scss (hunk 1 above)
Comment (shirokuronemuri) [reply frontend/src/components/common/header/header.module.scss]: fixed

File: frontend/src/components/common/header/header.tsx (hunk 2 above)
Comment (shirokuronemuri) [reply frontend/src/components/common/header/header.tsx]: fixed

File: frontend/src/components/common/header/header.tsx (hunk 3 above)
Comment (shirokuronemuri) [reply frontend/src/components/common/header/header.tsx]: fixed

File: frontend/src/components/common/header/header.tsx (hunk 4 above)
Comment (shirokuronemuri) [reply frontend/src/components/common/header/header.tsx]: couldn't make it work so I rewrote it in different way

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx (hunk 5 above)
Comment (shirokuronemuri) [reply frontend/src/components/notification-dropdown/components/notification/notification.tsx]: fixed

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx (hunk 6 above)
Comment (shirokuronemuri) [reply frontend/src/components/notification-dropdown/components/notification/notification.tsx]: rewritten

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx (hunk 7 above)
Comment (shirokuronemuri) [reply frontend/src/components/notification-dropdown/components/notification/notification.tsx]: done

File: frontend/src/components/notification-dropdown/components/notification/notification.tsx (hunk 8 above)
Comment (shirokuronemuri) [reply frontend/src/components/notification-dropdown/components/notification/notification.tsx]: done

**Comment by shirokuronemuri** [comment]
sorry for refactoring so much code ;;

**Review by so5tmaker — APPROVED** [review]

**Review by droztoff — APPROVED** [review]

**Review by kyrylo-burmelov — APPROVED** [review]
Really bad that you have so many updates that's not related to your task. It's hard to review PR. Next time please create separate tasks for huge refactoring updates that's not related to your task.

**Pull request merged**